crates/dbsp/src/operator/time_series/watermark.rs

```diff
@@ -76,8 +82,103 @@ where
     }
 }
 
+impl<B> Stream<RootCircuit, B>
+where
+    B: BatchReader + Clone + 'static,
+{
+    /// Computes the least upper bound over all records that occurred in the
+    /// stream with respect to some user-defined lattice.
+    ///
+    /// The primary use of this function is in time series analytics in
+    /// computing the largest timestamp observed in the stream, which can in
+    /// turn be used in computing retainment policies for data in this
+    /// stream and streams derived from it (see
+    /// [`Stream::integrate_trace_retain_keys`] and
+    /// [`Stream::integrate_trace_retain_values`]).
+    ///
+    /// Note: the notion of time here is distinct from the DBSP logical time and
+    /// represents one or several physical timestamps embedded in the input
+    /// data.
+    ///
+    /// In the special case where timestamps form a total order and the input
+    /// stream is indexed by time, the
+    /// [`watermark_monotonic`](`Stream::watermark_monotonic`) function can
+    /// be used instead of this method to compute the bound more
+    /// efficiently.
+    ///
+    /// # Arguments
+    ///
+    /// * `init` - initial value of the bound, usually the bottom element of the
```

Comment:
This is a problem; I am not sure it will always be obvious to the compiler how to choose this lower bound.
Could a None value be used instead?

Reply:
This makes me also wonder what will happen to nullable types. I guess the compiler will synthesize the right function...

Reply:
If the timestamp value has type `Option` then `None` can be used.  Otherwise, it should be the minimal value of the timestamp type. If it's a mutidimensional timestamp, the bottom element of the lattice is a tuple of minimal values for each dimension.